src/plots/cartesian/axes.js

```diff
@@ -1628,12 +1628,20 @@ axes.drawOne = function(gd, ax, opts) {
     var counterLetter = axes.counterLetter(axId);
     var mainSubplot = ax._mainSubplot;
     var mainLinePosition = ax._mainLinePosition;
+    var mainMirrorPosition = ax._mainMirrorPosition;
     var mainPlotinfo = fullLayout._plots[mainSubplot];
     var mainAxLayer = mainPlotinfo[axLetter + 'axislayer'];
     var subplotsWithAx = axes.getSubplots(gd, ax);
 
     var vals = ax._vals = axes.calcTicks(ax);
 
+    // Add a couple of axis properties that should cause us to recreate
+    // elements. Used in d3 data function.
+    var axInfo = [ax.mirror, mainLinePosition, mainMirrorPosition].join('_');
+    for(i = 0; i < vals.length; i++) {
+        vals[i].axInfo = axInfo;
+    }
```

Comment:
I think it was always incorrect in fact to have `makeDataFn` refer to axis attributes directly... they need to be in the data that's bound to the ticks by D3, otherwise changing them will not cause those ticks to be recreated. The right thing to do is what I do here: push the relevant axis attributes into the tick data, so the old ticks have the old axis attributes and the new ones have new attributes. Then `makeDataFn` goes away, it's just a static function `tickDataFn`.

I didn't add any new tests for this change, it was made in order to fix existing tests, which I guess were failing because we aren't doing as much full redrawing with the simplified pipeline.

Reply:
Nicely done. It might be nice to try to :hocho: this block

https://github.com/plotly/plotly.js/blob/3ea5c82de6790a3471fea10774fc50b9c3023ae6/src/plots/cartesian/axes.js#L1570-L1584

entirely in favor of better/more-general d3 data functions. I suspect this may lead to performance degradation on pan, so I would need to be careful.